packages/network-controller/tests/provider-api-tests/rpc-failover.ts

```diff
@@ -0,0 +1,504 @@
+              failoverComms.mockRpcCall({
+                request: requestToMock,
+                response: {
+                  result: 'ok',
+                },
+              });
+
+              const messenger = buildRootMessenger();
+
+              const result = await withNetworkClient(
+                {
+                  providerType,
+                  isRpcFailoverEnabled: true,
+                  failoverRpcUrls: ['https://failover.endpoint'],
+                  messenger,
+                  getRpcServiceOptions: (rpcEndpointUrl) => {
+                    const commonOptions = { fetch, btoa };
+                    // We need to return different results.
+                    // eslint-disable-next-line jest/no-conditional-in-test
+                    if (rpcEndpointUrl === 'https://failover.endpoint') {
+                      const headers: HeadersInit = {
+                        'X-Baz': 'Qux',
+                      };
+                      return {
+                        ...commonOptions,
+                        fetchOptions: {
+                          headers,
+                        },
+                      };
+                    }
+                    const headers: HeadersInit = {
+                      'X-Foo': 'Bar',
+                    };
+                    return {
+                      ...commonOptions,
+                      fetchOptions: {
+                        headers,
+                      },
+                      policyOptions: {
+                        backoff: new ConstantBackoff(backoffDuration),
+                        maxRetries: customMaxRetries,
+                        maxConsecutiveFailures: customMaxConsecutiveFailures,
+                      },
+                    };
+                  },
+                },
+                async ({ makeRpcCall, clock }) => {
+                  messenger.subscribe(
+                    'NetworkController:rpcEndpointRequestRetried',
+                    () => {
+                      // Ensure that we advance to the next RPC request
+                      // retry, not the next block tracker request.
+                      // We also don't need to await this, it just needs to
+                      // be added to the promise queue.
+                      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                      clock.tickAsync(backoffDuration);
+                    },
+                  );
+
+                  for (let i = 0; i < customNumRequestsToMake - 1; i++) {
+                    await ignoreRejection(makeRpcCall(request));
+                  }
+                  return await makeRpcCall(request);
+                },
+              );
+
+              expect(result).toBe('ok');
+            },
+          );
+        },
+      );
+    });
+  });
+
+  describe('if RPC failover functionality is not enabled', () => {
+    it(`throws even after ${maxConsecutiveFailures} unsuccessful attempts`, async () => {
```

Comment:
This is the new test being added to ensure that when disabling, even if failover URLs are specified, they do not get used.